lib/auth/authentication_backend.rb

```diff
@@ -1,20 +1,15 @@
 require_relative 'challenge_response'
 require_relative 'authenticated_response'
+require 'securerandom'
 
 module AuthenticationBackend
-  # Abstracted Exceptions
-  # Aws::CognitoIdentityProvider::Errors::NotAuthorizedException,
-  # Aws::CognitoIdentityProvider::Errors::UserNotFoundException,
-  # Aws::CognitoIdentityProvider::Errors::InvalidParameterException,
-  # Aws::CognitoIdentityProvider::Errors::CodeMismatchException
-  # Aws::CognitoIdentityProvider::Errors::ServiceError
-  # Aws::CognitoIdentityProvider::Errors::AliasExistsException,
-  # Aws::CognitoIdentityProvider::Errors::UsernameExistsException => e
-  # Aws::CognitoIdentityProvider::Errors::ResourceNotFoundException
   class NotAuthorizedException < StandardError; end
```

Comment:
I like how you've done this; really good stuff so far. Minor observation, it seems the convention in ruby with custom exception is to end their names in Error so maybe
 ```
class UserGroupNotFoundError < StandardError; end
 class AuthenticationBackendError < StandardError; end	 
 class AuthenticationBackendError < StandardError; end
 class UsernameExistsError < StandardError; end
 class GroupExistsError < StandardError; end
```